cookbooks/travis_build_environment/recipes/python.rb

```diff
@@ -0,0 +1,175 @@
+    )
+  end
+
+  venv_fullname = "#{virtualenv_root}/#{pyname}"
+
+  remote_file downloaded_tarball do
+    source ::File.join(
+      'https://s3.amazonaws.com/travis-python-archives/binaries',
+      node['platform'],
+      node['platform_version'],
+      node['kernel']['machine'],
+      ::File.basename(downloaded_tarball)
+    )
+    owner 'root'
+    group 'root'
+    mode 0o644
+    ignore_failure true
+  end
+
+  bash "extract #{downloaded_tarball}" do
+    code "tar -xjf #{downloaded_tarball} --directory /"
+    creates "/opt/python/#{py}"
+    environment build_environment
+    only_if { File.exist?(downloaded_tarball) }
+  end
+
+  bash "build #{py}" do
+    code "python-build #{py} /opt/python/#{py}"
+    creates "/opt/python/#{py}"
+    environment build_environment
+    not_if { File.exist?("/opt/python/#{py}") }
+  end
+
+  link "/opt/python/#{py}/bin/#{pyname}" do
+    to "/opt/python/#{py}/bin/python"
+    owner node['travis_build_environment']['user']
+    group node['travis_build_environment']['group']
+  end
+
+  bindirs << "/opt/python/#{py}/bin"
+
+  bash "create virtualenv at #{venv_fullname} from #{py}" do
+  code "virtualenv --python=/opt/python/#{py}/bin/python #{venv_fullname}"
+  owner node['travis_build_environment']['user']
+  group node['travis_build_environment']['group']
+  end
+
+  node['travis_build_environment']['pyenv']['aliases'].fetch(py, []).each do |pyalias|
+    if /^\d+\.\d+(?:\.\d+)?(?:-dev)?$/ =~ py
+      pyaliasname = "python#{pyalias}"
+    else
+      pyaliasname = pyalias
+    end
+
+    link "#{virtualenv_root}/#{pyaliasname}" do
+      to venv_fullname
+      owner node['travis_build_environment']['user']
+      group node['travis_build_environment']['group']
+    end
+  end
+
+  if py =~ /^pypy3/
+    link "/opt/python/#{py}/bin/pypy3" do
+      to "/opt/python/#{py}/bin/#{py}"
+      not_if "test -f /opt/python/#{py}/bin/pypy3"
+    end
+  end
+
+  packages = []
+
+  node['travis_build_environment']['pyenv']['aliases'].fetch(py, []).concat(['default', py]).each do |name|
+    packages.concat(node['travis_build_environment']['pip']['packages'].fetch(name, []))
+  end
+
+  execute "install wheel in #{py}" do
+    command "#{venv_fullname}/bin/pip install --upgrade wheel"
+    user node['travis_build_environment']['user']
+    group node['travis_build_environment']['group']
+    environment(
+      'HOME' => node['travis_build_environment']['home']
+    )
+  end
+
+  execute "install packages in #{py}" do
+    command "#{venv_fullname}/bin/pip install --upgrade #{packages.join(' ')}"
+    user node['travis_build_environment']['user']
+    group node['travis_build_environment']['group']
+    environment(
+      'HOME' => node['travis_build_environment']['home']
+    )
+  end
+end
+
+template '/etc/profile.d/pyenv.sh' do
+  source 'pyenv.sh.erb'
+    owner node['travis_build_environment']['user']
+    group node['travis_build_environment']['group']
+    mode 0o644
+    variables(
```

Comment:
So close again.  The `source`, `owner`, `group`, `mode`, `variables`, and `backup` calls should all be aligned one indentation level in from the outer `template` resource.  Rubocop auto fix should handle this for you ❤️